### PR TITLE
feat!: add support for preserving and optionally applying guardrails to reasoning traces

### PR DESCRIPTION
**Pull request by Pouyanpi**
## Summary Generated by Copilot

This pull request introduces enhancements to the LLM generation pipeline by adding support for reasoning traces and improving the handling of parsed task outputs. Key changes include the introduction of new utility functions, updates to existing methods to process reasoning traces, and modifications to ensure consistent handling of parsed outputs across various LLM tasks.

### Enhancements to Reasoning Trace Support:
* Added a new context variable `reasoning_trace_var` to store reasoning traces and utility functions such as `_record_reasoning_trace`, `_assemble_response`, and `_process_parsed_output` to manage reasoning traces and integrate them into responses. (`nemoguardrails/context.py` [[1]](diffhunk://#diff-43d1ee5dfe7869aeaf94a061429ce13a0b0ef7cc64e44dbf7c94c39d0e00de4cR36-R39) `nemoguardrails/actions/llm/generation.py` [[2]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R1405-R1428)
* Introduced `_get_apply_to_reasoning_traces` to fetch configuration values for including reasoning traces in outputs, and `_include_reasoning_traces` to encapsulate this logic. (`nemoguardrails/actions/llm/generation.py` [nemoguardrails/actions/llm/generation.pyR1384-R1387](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R1384-R1387))

### Consistent Handling of Parsed Outputs:
* Updated multiple methods (e.g., `generate_user_intent`, `generate_bot_message`, `generate_value`) to extract and use the `text` property from `ParsedTaskOutput` objects for consistency. (`nemoguardrails/actions/llm/generation.py` [[1]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R446) [[2]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R658) [[3]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R1076)
* Added reasoning trace processing in key generation methods like `generate_bot_message` and `generate_intent_steps_message`. (`nemoguardrails/actions/llm/generation.py` [[1]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R913-R916) [[2]](diffhunk://#diff-30aeca7bd0b244d8479885d39179c92887233805f4e649e59b9a4548b6b24672R1371)

### Utility and Structural Improvements:
* Introduced `ReasoningExtractionResult` as a `dataclass` to encapsulate cleaned text and reasoning traces extracted from LLM outputs. (`nemoguardrails/llm/filters.py` [nemoguardrails/llm/filters.pyL18-R36](diffhunk://#diff-bdcdcf1c11cae24a95259ec758586747b2e95fa7f1274bac9b780f1d07a981b7L18-R36))
* Added `get_and_clear_reasoning_trace` to retrieve and clear reasoning traces from the context. (`nemoguardrails/actions/llm/utils.py` [nemoguardrails/actions/llm/utils.pyR580-R591](diffhunk://#diff-842f13dbe221e5243ab5a8e366c77e373c5038cb5640584593ec0f8bc6041136R580-R591))

### Updates to Other LLM Modules:
* Applied similar text extraction logic to other modules, including content safety and self-check actions, ensuring uniformity in handling parsed outputs. (`nemoguardrails/library/content_safety/actions.py` [[1]](diffhunk://#diff-525819b9fd67ffe6d02f353cc52675c791a24fa128fc4e6d40c25adf06df1d29R83) `nemoguardrails/library/self_check/facts/actions.py` [[2]](diffhunk://#diff-c2d4db077638c80ccb327834c44d255ba645f35f70f63f234d112e1c4ee033acR85)

These changes collectively enhance the flexibility and traceability of LLM outputs, improving both debugging capabilities and the overall user experience.

-----------------------

## Impact

tldr; 
**this feature ensures that reasoning traces are always returned to the user. Removing it would affect internals of nemoguardrails, but not the output that users receive.**

**Interaction between apply_to_reasoning_traces and remove_thinking_traces:**

* extraction always happens: reasoning traces are always extracted from LLM responses
* storage always happens: extracted traces are always stored in `reasoning_trace_var`

**conditional response inclusion:**

* when apply_to_reasoning_traces is TRUE: Traces are prepended to visible output
* when apply_to_reasoning_traces is FALSE: Traces are not included in visible output

**visible vs. stored output:**

* the traces are always part of what's returned to the user interface
* but they are NOT stored in the $bot_message context when the setting is FALSE

### implementation details:

_assemble_response(text, trace, include_reasoning) controls this behavior
Returns trace + text when include_reasoning is TRUE
Returns just text when include_reasoning is FALSE
Flow control: The decision happens in _process_parsed_output() where both storage and assembly occur

-  [x] breaking change?
- [x] push llm reasoning output rails test
- [ ] do same for colang 2

requires #1161

### CODE REVIEW

**Comment by Pouyanpi** [comment]
I feel uneasy about the way `nemoguardrails.llm.generation` does:

```python
result = self.llm_task_manager.parse_task_output(…)
text   = result.text # this is new before `result` was plain str
# …then all kinds of ad-hoc cleaning: .strip(), strip_quotes(), 
# get_first_nonempty_line(), get_multiline_response(), split("\n"), etc.
```

all over the place.  We end up with:

- Duplication: every branch re-implements the same little clean-up steps (strip whitespace, remove quotes, pick line 1 vs N, etc.).

- Inconsistency: sometimes we strip before removing quotes, sometimes after; sometimes we drop blank lines early, sometimes late.

- Leaky abstractions: our action code has to know the gory details of how each prompts output is formatted, instead of just “I asked for a canonical intent, give me one.”

- Type confusion: it mutates or shadows result (a ParsedTaskOutput) into a plain str, losing the .reasoning_trace and obligation to always deal with the richer type.


We MUST stop inlining dozens of little string hacks in every action.

**Comment by Pouyanpi** [comment]
This module `generation.py` is doing too much, with tangled responsibilities, ad-hoc hacks, and barely controlled state.  Splitting it up, locking down state flows, adding types/tests, and centralizing ouar LLM invocation + post-processing logic will dramatically improve clarity, safety, and maintainability.

File: nemoguardrails/actions/llm/utils.py [hunk 1]
```diff
@@ -577,3 +577,15 @@ def escape_flow_name(name: str) -> str:
     # removes non-word chars and leading digits in a word
     result = re.sub(r"\b\d+|[^\w\s]", "", result)
     return result
+
+
+def get_and_clear_reasoning_trace() -> Optional[str]:
```
Comment (Pouyanpi) [inline nemoguardrails/actions/llm/utils.py]:
```python
get_and_clear_reasoning_trace_contextvar
```

File: tests/test_llmrails_reasoning.py [hunk 2]
```diff
@@ -92,5 +92,5 @@ async def compute(what: Optional[str] = "2 + 3"):
     bot_message = await llm_rails.generate_async(messages=messages)
     assert bot_message == {
         "role": "assistant",
-        "content": "The answer is 5\nAre you happy with the result?",
+        "content": "<think>some important COT text</think>The answer is 5\nAre you happy with the result?",
```
Comment (Pouyanpi) [inline tests/test_llmrails_reasoning.py:95]: This is the consequence of the requested feature, we always include reasoning traces in the generation.

Comment (Pouyanpi) [reply tests/test_llmrails_reasoning.py:95]: @trebedea for viz.

File: nemoguardrails/llm/taskmanager.py [hunk 3]
```diff
@@ -52,13 +53,68 @@
 from nemoguardrails.rails.llm.config import MessageTemplate, RailsConfig
 
 
+def output_has_reasoning_traces(output: str, start_token: str, end_token: str) -> bool:
+    """Checks if the output string contains both start and end reasoning tokens."""
+    return start_token in output and end_token in output
```
Comment (cparisien) [inline nemoguardrails/llm/taskmanager.py:58]: Would this break on text that happens to have multiple "<think>" strings, possibly out of the expected order?

File: nemoguardrails/rails/llm/config.py [hunk 4]
```diff
@@ -1199,11 +1205,10 @@ def check_reasoning_traces_with_dialog_rails(cls, values):
...
         # dialog rails are activated (explicitly or implicitly)
-        has_dialog_rails = (
-            bool(dialog_rails)
-            or bool(values.get("user_messages"))
-            or bool(values.get("bot_messages"))
-            or bool(values.get("flows"))
+        has_dialog_rails = bool(dialog_rails) or (
+            bool(values.get("user_messages"))
+            and bool(values.get("bot_messages"))
+            # and bool(values.get("flows"))
```
Comment (cparisien) [inline nemoguardrails/rails/llm/config.py]: Why is this commented out now?

File: tests/test_filters.py [hunk 5]
```diff
@@ -92,29 +94,141 @@ def test_last_turns():
...
+            result.append(part)
+        elif isinstance(part, tuple) and len(part) == 2:
+            char, count = part
+            result.append(char * count)
+        else:
+            raise TypeError(f"Invalid part type in _build_test_string: {part}")
+    return "".join(result)
+
+
 @pytest.mark.parametrize(
```
Comment (cparisien) [inline tests/test_filters.py:116]: Also add out-of-order cases like "[END] [START] [START]"

**Review by cparisien — APPROVED** [review]
I'm good with this, just add some more tests for unexpected patterns that might look like reasoning traces but aren't.

File: nemoguardrails/llm/taskmanager.py (hunk 3 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/taskmanager.py:58]: no it shouldn't 👍🏻

File: nemoguardrails/rails/llm/config.py (hunk 4 above)
Comment (Pouyanpi) [reply nemoguardrails/rails/llm/config.py]: it is being discussed in #1161

File: nemoguardrails/context.py [hunk 6]
```diff
@@ -32,3 +33,7 @@
 # The raw LLM request that comes from the user.
 # This is used in passthrough mode.
 raw_llm_request = contextvars.ContextVar("raw_llm_request", default=None)
+
+reasoning_trace_var: ContextVar[str] = contextvars.ContextVar(
```
Comment (trebedea) [inline nemoguardrails/context.py]:
```suggestion
reasoning_trace_var: ContextVar[Optional[str]] = ContextVar(
```

Comment (trebedea) [reply nemoguardrails/context.py]: Replace `contextvars.ContextVar` with `ContextVar` and also add type hints for the other vars as well.

File: tests/test_filters.py (hunk 5 above)
Comment (Pouyanpi) [reply tests/test_filters.py:116]: Totally, just added 👍🏻

File: nemoguardrails/llm/filters.py [hunk 7]
```diff
@@ -15,14 +15,25 @@
...
 from nemoguardrails.actions.llm.utils import (
     get_colang_history,
     remove_action_intent_identifiers,
 )
 
 
+@dataclass
+class ReasoningExtractionResult:
+    """
+    Holds cleaned text and optional chain-of-thought reasoning trace extracted from LLM output.
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py]:
```suggestion
    Holds cleaned response text and optional chain-of-thought reasoning trace extracted from LLM output.
```

File: nemoguardrails/llm/filters.py [hunk 8]
```diff
@@ -15,14 +15,25 @@
...
+    """
+
+    text: str
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:33]: One comment here, in some cases where the `max_length` is too small and the reasoning is long, there might not be any actual response text. There are also (rare) cases of infinite reasoning traces. Should we treat these situations?

File: nemoguardrails/llm/filters.py [hunk 9]
```diff
@@ -15,14 +15,25 @@
...
+from typing import List, Optional, Tuple
 
 from nemoguardrails.actions.llm.utils import (
     get_colang_history,
     remove_action_intent_identifiers,
 )
 
 
+@dataclass
+class ReasoningExtractionResult:
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:28]: Should we also keep the original LLM response? Probably we do not need it with the current implementation, but we might use it sometime in the future.

File: nemoguardrails/llm/filters.py [hunk 10]
```diff
@@ -439,24 +450,94 @@ def conversation_to_events(conversation: List) -> List[dict]:
     return events
 
 
-def remove_reasoning_traces(response: str, start_token: str, end_token: str) -> str:
-    """Removes the text between the first occurrence of the start token and the
-    last occurrence of the last token, if these tokens exist in the response.
+def _find_token_positions_for_removal(
+    response: str, start_token: str, end_token: str
+) -> Tuple[int, int]:
+    """Helper function to find token positions specifically for text removal.
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:456]:
```suggestion
    """Helper function to find token positions specifically for text removal. This is useful, for example, to remove reasoning traces from a reasoning LLM response.
```

File: nemoguardrails/llm/filters.py [hunk 11]
```diff
@@ -439,24 +450,94 @@ def conversation_to_events(conversation: List) -> List[dict]:
...
+
+    Args:
+        response(str): The text to search in
+        start_token(str): The token marking the start of text to remove
+        end_token(str): The token marking the end of text to remove
 
-    This utility function is useful to strip reasoning traces from reasoning LLMs
-    that encode the reasoning traces between specific tokens.
+    Returns:
+        A tuple of (start_index, end_index) where both are -1 if not found
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py]:
```suggestion
        A tuple of (start_index, end_index) marking the span to remove; both indices are -1 if start_token and end_token are not provided.
```

File: nemoguardrails/llm/filters.py [hunk 12]
```diff
@@ -439,24 +450,94 @@ def conversation_to_events(conversation: List) -> List[dict]:
...
+    if not start_token or not end_token:
+        return -1, -1
+
+    start_index = response.find(start_token)
+    # if the start index is missing, this is probably a continuation of a bot message
+    # started in the prompt.
+    if start_index == -1:
+        start_index = 0
+
+    end_index = response.rfind(end_token)
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:483]: What happens if the `end_token` is missing (due to small `max_length` or infinite reasoning loop)? We should take this into consideration and document it.

File: nemoguardrails/llm/filters.py [hunk 13]
```diff
@@ -439,24 +450,94 @@ def conversation_to_events(conversation: List) -> List[dict]:
...
-            start_index = 0
-        end_index = response.rfind(end_token)
-        if end_index == -1:
-            return response
-
-        if start_index != -1 and end_index != -1 and start_index < end_index:
-            return response[:start_index] + response[end_index + len(end_token) :]
-
-    return response
+    if not start_token or not end_token:
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:474]: Should we put `start_token: Optional[str]` in the method definition? Same for `end_token`?

File: nemoguardrails/llm/filters.py [hunk 14]
```diff
@@ -439,24 +450,94 @@ def conversation_to_events(conversation: List) -> List[dict]:
...
+        without reasoning traces and the extracted reasoning trace, if any.
+    """
+    if not start_token or not end_token:
+        return ReasoningExtractionResult(text=response, reasoning_trace=None)
+
+    start_index, end_index = find_reasoning_tokens_position(
+        response, start_token, end_token
+    )
+
+    if end_index == -1:
```
Comment (trebedea) [inline nemoguardrails/llm/filters.py:537]: If there is a `start_index` and missing `end_index` we should consider that there is no actual response and everything is a reasoning trace.

File: nemoguardrails/llm/taskmanager.py [hunk 15]
```diff
@@ -52,13 +53,62 @@
...
+
+
+@dataclass
+class ParsedTaskOutput:
```
Comment (trebedea) [inline nemoguardrails/llm/taskmanager.py:62]: This class seems similar to `ReasoningExtractionResult`. Does it make sense to have both?

File: nemoguardrails/llm/taskmanager.py [hunk 16]
```diff
@@ -52,13 +53,62 @@
...
+
+
+def get_reasoning_token_tags(config, task):
+    model = get_task_model(config, task)
+
+    if model and model.reasoning_config:
+        start_token = model.reasoning_config.start_token
+        end_token = model.reasoning_config.end_token
+    else:
+        start_token = ""
```
Comment (trebedea) [inline nemoguardrails/llm/taskmanager.py]: Should these be `None` if `model.reasoning_config` is not defined?

File: nemoguardrails/llm/taskmanager.py [hunk 17]
```diff
@@ -78,7 +128,7 @@ def __init__(self, config: RailsConfig):
         self.env.filters["to_chat_messages"] = to_chat_messages
         self.env.filters["verbose_v1"] = verbose_v1
 
-        self.output_parsers = {
+        self.output_parsers: Dict[Optional[str], Callable] = {
```
Comment (trebedea) [inline nemoguardrails/llm/taskmanager.py:131]:
```suggestion
        self.output_parsers: Dict[str, Callable] = {
```

File: nemoguardrails/llm/taskmanager.py [hunk 18]
```diff
@@ -308,36 +358,35 @@ def render_task_prompt(
 
     def parse_task_output(
         self, task: Task, output: str, forced_output_parser: Optional[str] = None
-    ):
-        """Parses the output for the provided tasks.
+    ) -> ParsedTaskOutput:
+        """ """
```
Comment (trebedea) [inline nemoguardrails/llm/taskmanager.py]: Missing docstring.

File: nemoguardrails/llm/taskmanager.py [hunk 19]
```diff
@@ -308,36 +358,35 @@ def render_task_prompt(
...
-        ):
-            start_token = model.reasoning_config.start_token
-            end_token = model.reasoning_config.end_token
-            output = remove_reasoning_traces(output, start_token, end_token)
-
-        if output_parser:
-            return output_parser(output)
-        else:
-            return output
+        return ParsedTaskOutput(text=parsed_text, reasoning_trace=reasoning_trace)
```
Comment (trebedea) [inline nemoguardrails/llm/taskmanager.py:404]: So actually `ParsedTaskOutput.text` can be either the response stripped of reasoning traces or the response including reasoning traces. We should mention this in the docstrings (and documentation), no?

File: nemoguardrails/rails/llm/config.py [hunk 20]
```diff
@@ -444,6 +444,12 @@ class OutputRails(BaseModel):
         description="Configuration for streaming output rails.",
     )
 
+    apply_to_reasoning_traces: bool = Field(
+        default=False,
+        description="If True, output rails will apply guardrails to both reasoning traces and output response. "
+        "If False, output rails will only apply guardrails to the output response while keeping reasoning traces unaltered. ",
```
Comment (trebedea) [inline nemoguardrails/rails/llm/config.py]:
```suggestion
        "If False, output rails will only apply guardrails to the output response excluding the reasoning traces, thus keeping reasoning traces unaltered. ",
```

File: tests/test_llmrails_reasoning.py (hunk 2 above)
Comment (trebedea) [reply tests/test_llmrails_reasoning.py:95]: OK, this looks good. 
But we are stripping the reasoning traces when we construct the chat history?

File: nemoguardrails/actions/v2_x/generation.py [hunk 21]
```diff
@@ -302,6 +302,8 @@ async def generate_user_intent(
             Task.GENERATE_USER_INTENT_FROM_USER_ACTION, output=result
         )
 
+        result = result.text
```
Comment (trebedea) [inline nemoguardrails/actions/v2_x/generation.py:305]: Are we supporting reasoning traces for Colang 2.0? The logic here seems much simpler that for Colang 1.0 and there is no reasoning traces logic.

File: tests/test_llmrails_reasoning_output_rails.py [hunk 22]
```diff
@@ -0,0 +1,312 @@
...
+        """,
+    )
+
+
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    "test_case",
+    [
+        ReasoningTraceTestCase(
+            description="Remove thinking traces and show error when guardrail is enabled",
```
Comment (trebedea) [inline tests/test_llmrails_reasoning_output_rails.py:98]:
```suggestion
            description="Remove reasoning traces and show error when rails are enabled to be applied on the reasoning traces",
```

Is this the intent?

File: tests/test_llmrails_reasoning_output_rails.py [hunk 23]
```diff
@@ -0,0 +1,312 @@
...
+        ),
+        ReasoningTraceTestCase(
+            description="Preserve thinking traces and show error when guardrail is enabled",
+            remove_thinking_traces=False,
+            apply_to_reasoning_traces=True,
+            expected_think_tag=True,
+            expected_error_message=True,
+        ),
+        ReasoningTraceTestCase(
+            description="Remove thinking traces and show error when both flags are disabled",
```
Comment (trebedea) [inline tests/test_llmrails_reasoning_output_rails.py:119]: "thinking traces" => "reasoning traces" all over.

File: tests/test_llmrails_reasoning_output_rails.py [hunk 24]
```diff
@@ -0,0 +1,312 @@
...
+
+    1. When remove_thinking_traces=True:
+       - The model is configured to remove thinking traces
+       - However, the actual removal depends on apply_to_reasoning_traces
+
+    2. When apply_to_reasoning_traces=True:
+       - The output rail will check for and report think tags
+       - Because we expect the think tag to be present as output rails explicitly requires it
+
+    3. When apply_to_traces=False:
```
Comment (trebedea) [inline tests/test_llmrails_reasoning_output_rails.py]: This be `apply_to_reasoning_traces`, no? It appears a couple of times like this.

**Review by trebedea — CHANGES_REQUESTED** [review]
The heavy lifting in this MR looks good. There are some minor fixes, but I have not seen anything major that needs to be improved. Nice job!

**Comment by trebedea** [comment]
> This module `generation.py` is doing too much, with tangled responsibilities, ad-hoc hacks, and barely controlled state. Splitting it up, locking down state flows, adding types/tests, and centralizing ouar LLM invocation + post-processing logic will dramatically improve clarity, safety, and maintainability.

Completely agreed! There is a lot of technical debt: complex logic, ad-hoc fixes, and duplication in `generation.py`.
We just kept adding code to make things work, but it should be time to restructure it. Maybe in next release?

File: tests/test_llmrails_reasoning_output_rails.py (hunk 24 above)
Comment (Pouyanpi) [reply tests/test_llmrails_reasoning_output_rails.py]: yes thanks for catching this 👍🏻

File: nemoguardrails/actions/v2_x/generation.py (hunk 21 above)
Comment (Pouyanpi) [reply nemoguardrails/actions/v2_x/generation.py:305]: it is in my todos, if time permits

File: tests/test_llmrails_reasoning_output_rails.py (hunk 22 above)
Comment (Pouyanpi) [reply tests/test_llmrails_reasoning_output_rails.py:98]: Yes exactly, I have a separate branch that I will push once this is merged, there I'm deprecating `remove_thinking_traces` and will adapt the tests accordingly. Yes you got the intent right, TY!

File: nemoguardrails/llm/filters.py (hunk 12 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/filters.py:483]: Oh I just recalled (QA's), good catch 👍🏻 

somehow documented in the `test_filters.py`

```python
        (
            "Only [START] start token",
            "[START]",
            "[END]",
            "Only [START] start token",
            None,
        ),

```

If we should do more, let's do it in a separate PR/issue.

File: nemoguardrails/llm/filters.py (hunk 13 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/filters.py:474]: Yes!

File: nemoguardrails/llm/taskmanager.py (hunk 15 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/taskmanager.py:62]: I think it does, they have a different intent and their similarity is a coincidence now 👍🏻

File: nemoguardrails/llm/taskmanager.py (hunk 17 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/taskmanager.py:131]: it should be optional given the existing logic, no?

```python
        parser_fn = self.output_parsers.get(parser_name)

        if parser_fn:
            parsed_text = parser_fn(output)
        else:
            logging.info("No output parser found for %s", prompt.output_parser)
            parsed_text = output

```

File: nemoguardrails/context.py (hunk 6 above)
Comment (Pouyanpi) [reply nemoguardrails/context.py]: will do for the rest in a separate PR, otherwise codecov expects me to have all of them tested in this PR 👍🏻

File: tests/test_llmrails_reasoning.py (hunk 2 above)
Comment (Pouyanpi) [reply tests/test_llmrails_reasoning.py:95]: Will be part of a subsequent PR, and it will only strip the tokens `history` in `_render_string` method in `llmtaskmanager.py`.

File: nemoguardrails/llm/filters.py (hunk 8 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/filters.py:33]: did I address it in another comment? 

Let's address it in a separate PR if time permits, I understand this edge case and probably we should handle it.

File: tests/test_llmrails_reasoning.py (hunk 2 above)
Comment (Pouyanpi) [reply tests/test_llmrails_reasoning.py:95]: #1169

**Comment by Pouyanpi** [comment]
merged before: #1169

File: nemoguardrails/llm/filters.py (hunk 12 above)
Comment (trebedea) [reply nemoguardrails/llm/filters.py:483]: Ok, sounds good.

File: nemoguardrails/llm/taskmanager.py (hunk 17 above)
Comment (trebedea) [reply nemoguardrails/llm/taskmanager.py:131]: Not sure I follow. The dict always has a string as a key (the task name), but there are some tasks that do not have any output parser assigned. Maybe we want to say something like : 

`self.output_parsers: Dict[str, Optional[Callable]] ` ?

Just to highlight that some tasks might not have any output parser?

Comment (Pouyanpi) [reply nemoguardrails/llm/taskmanager.py:131]: It can be that the output_parser is not defined for a task in a given prompt (user provided),

as it is optional str in `MessageTemplate` class:
```python
    output_parser: Optional[str] = Field(
        default=None,
        description="The name of the output parser to use for this prompt.",
    )
```

and previously we were doing

```python
        output_parser = None
        if forced_output_parser:
            output_parser = self.output_parsers.get(forced_output_parser)
        elif prompt.output_parser:
            output_parser = self.output_parsers.get(prompt.output_parser)
        if not output_parser:
            logging.info("No output parser found for %s", prompt.output_parser)

```

also we are not raising any sort of error as currently it is not required to define a parser per task, for example see [self check output action](https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/nemoguardrails/library/self_check/output_check/actions.py#L82-L88).

File: nemoguardrails/llm/filters.py (hunk 14 above)
Comment (Pouyanpi) [reply nemoguardrails/llm/filters.py:537]: shall we raise an error?

Guardrails won't work without an actual response.

Comment (Pouyanpi) [reply nemoguardrails/llm/filters.py:537]: I'll merge this PR as it is getting very large. I'll open a separate PR to address this issue, I'll add you as a reviewer.

Comment (trebedea) [reply nemoguardrails/llm/filters.py:537]: Yes, let's merge this PR. It's already a lot here.
